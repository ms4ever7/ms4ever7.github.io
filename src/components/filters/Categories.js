import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Collapsible from 'react-collapsible';

import { selectCategory } from '../../redux/actions/filters';

const DEFAULT_CLASS_NAME = 'categories__list-item';
const CATEGORIES_LIST = ['Athletic', 'Fashion Sneakers', 'Boots', 'Oxfords'];

class Categories extends Component {
  static propTypes = {
    selectedCategories: PropTypes.array,
    selectCategory: PropTypes.func
  }

  static defaultProps = {
    selectedCategories: [],
    selectCategory: () => {}
  }

  state = {
    categoriesOpened: false
  }

  toggleCategory = (category) => {
    this.props.selectCategory(category);
  }

  get triggerElement() {
    const { categoriesOpened } = this.state;
    const iconToShow = categoriesOpened ? '-' : '+';

    return <div className='categories__wrapper--trigger layout layout-align-space-between-end'>
      <div className='categories__list-title'>Categories:</div>
      <div>{iconToShow}</div>
    </div>
  }

  get categories() {
    const { selectedCategories } = this.props;
    
    return <Collapsible 
      trigger={this.triggerElement}
      onOpen={() => this.setState({ categoriesOpened: true })} 
      onClose={() => this.setState({ categoriesOpened: false })}
      >
      {CATEGORIES_LIST.map((category, index) => {
        let className = DEFAULT_CLASS_NAME;

        if (selectedCategories && selectedCategories.length && selectedCategories.includes(category)) {
          className = `${DEFAULT_CLASS_NAME} list-item--selected`
        }

        return <div
          key={index}
          className={className}
          onClick={() => this.toggleCategory(category)}>
          {category}
        </div>
      })}
    </Collapsible>
  }

  render() {
    return (
      <div className='categories__wrapper'>
        {this.categories}
      </div>
    );
  }
}

const mapStateToProps = ({ filters }) => ({
  selectedCategories: filters.selectedCategories
});

const mapDispatchToProps = dispatch => ({
  selectCategory: category => dispatch(selectCategory(category))
});


export default connect(mapStateToProps, mapDispatchToProps)(Categories);