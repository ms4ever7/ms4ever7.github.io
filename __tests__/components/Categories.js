import React from 'react';
import configureMockStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Categories from '../../src/components/filters/Categories';

describe('Categories', () => {
  let wrapper;
  let mockStore;

  beforeEach(() => {
    mockStore = configureMockStore([]);
  });

  it('should render correctly', () => {
    const props = fulfillProps();
    wrapper = shallow(
      <Categories />,
      {
        context: { store: mockStore(props) }
      }
    );

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});


const fulfillProps = (props = {}) => ({
  products: {
    items: [],
    ...props
  },
  filters: {
    selectCategory: [],
    selectColors: []
  }
});
