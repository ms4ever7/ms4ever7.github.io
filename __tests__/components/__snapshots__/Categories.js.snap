// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories should render correctly 1`] = `
<Categories
  selectCategory={[Function]}
  selectedCategories={Array []}
/>
`;
